### PR TITLE
Payment methods for billing address

### LINKED ISSUE #2878: Get the payment methods that are available for specific country.
## Steps to reproduce the issue
0. Check Default Country in Magento panel
1. In Magento set one of Payment Methods for Specific Country, for example:
**Check/Money Order** → Payment from Specific Countries: **Albania**
(Rest of payment methods should be available for all countries) 
2. Add any product to cart and go to checkout
3. Fill in all required fields and in Payment section choose Country: Albania
4. Look at available Payment methods

## Current behavior
There are available only these payment methods, which are refer to Default Country

## Expected behavior
Payment methods should depend on chosen country.

## Can you handle fixing this bug by yourself?

- [ ] YES
- [x] NO

## Which [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html) state this refers to? Info for developer.
Pick one option.

- [ ] This is a bug report for test version on https://test.storefrontcloud.io - In this case Developer should create branch from `develop` branch and create Pull Request `2. Feature / Improvement` back to `develop`.
- [ ] This is a bug report for current Release Candidate version on https://next.storefrontcloud.io - In this case Developer should create branch from `release` branch and create Pull Request `3. Stabilisation fix` back to `release`.
- [ ] This is a bug report for current Stable version on https://demo.storefrontcloud.io and should be placed in next stable version hotfix - In this case Developer should create branch from `hotfix` or `master` branch and create Pull Request `4. Hotfix` back to `hotfix`.

## Environment details
- Browser: Chrome 73.0.3683.103
- OS: Windows 10 Pro
- Code Version: v1.9.0

### PR DESCRIPTION
**Pull request by rain2o**
### Related issues
closes #2878 

### Short description and why it's useful
Started this at the Paris hackathon but had to finish it up today.

In Magento at least, not sure about other backend platforms, you have the ability to restrict payment methods per country, not just per store view. There was already an API call being made to `/shipping-information` which returns the allowed payment methods for the data given. This PR utilized this existing integration to do a few things.

1. Add billing address data to request data if it's available.
2. When it's called for updating cart price, go ahead and save the results if there are any since we have them.
3. Update `syncPaymentMethods` to check if we have billing address data and use this endpoint for more accurate results.
4. Call `syncPaymentMethods` when the billing address country is changed.

### Which environment this relates to
Check your case. In case of any doubts please read about [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html)

- [x] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [ ] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [ ] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

### Upgrade Notes and Changelog

- [x] No upgrade steps required (100% backward compatibility and no breaking changes)
- [ ] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/docs/guide/upgrade-notes/README.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing VS sites with this new feature

**IMPORTANT NOTICE** - Remember to update `CHANGELOG.md` with description of your change

### Contribution and currently important rules acceptance

- [x] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: core/modules/checkout/components/Payment.ts [hunk 1]
```diff
@@ -233,6 +233,9 @@ export const Payment = {
 
       // Let anyone listening know that we've changed payment method, usually a payment extension.
       this.$bus.$emit('checkout-payment-method-changed', this.payment.paymentMethod)
+    },
+    changeCountry () {
+      this.$bus.$emit('checkout-before-paymentMethods')
```
Comment (andrzejewsky) [inline core/modules/checkout/components/Payment.ts]:
```suggestion
      this.$store.dispatch('cart/syncPaymentMethods', { forceServerSync: true })
```

**Comment by andrzejewsky** [comment]
just noticed you @patzick wrote same comment 😄 @rain2o please apply these changes and we can merge it, good work 👍

File: core/modules/cart/types/BillingAddress.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+export default interface ShippingAddress {
```
[the remark this answers is not in the document]
Comment (rain2o) [reply core/modules/cart/types/BillingAddress.ts]: Ah, complete oversight on my part. It's been updated.

File: core/modules/checkout/components/Payment.ts (hunk 1 above)
Comment (rain2o) [reply core/modules/checkout/components/Payment.ts]: Ok thanks, I forgot about the move away from event bus for things like this. Updated.

**Comment by andrzejewsky** [comment]
@rain2o please look at tests, there is one failed, but should be easy to fix.

**Comment by rain2o** [comment]
@andrzejewsky Tests updated and passed.

**Comment by ArturDivante** [comment]
@rain2o @andrzejewsky 
I have a problem verifying this functionality.

When I set given payment method (in my example - check/money order)  "for all countries", it's correctly returned in payment-methods request response, among others.

However, when I set it only for specific countries, (tried it with many diffrent ones), it's not available. I tried to set countires from shipping section as well as from payments section. Every timie i make such change, new payment-methods request is send, but in repsonse there are only "global" methods, set for all countries. Any ideas?

**Comment by rain2o** [comment]
@ArturDivante That is strange, I can't replicate that on my local. The payment methods should return based on the country filled out in the billing address. Each time you switch the country value on the Billing address section it should make a new request with that country. And you're saying when you do this for one of the specified countries of the payment method, it isn't returning that specific payment method?

I just tested against the demo Magento site and I am seeing Check / Money Order show up only when I select United States as my billing country.

**Comment by rain2o** [comment]
@ArturDivante This is what I get when connected to http://demo-magento2.vuestorefront.io on my local.
vsf-payment-methods-country

**Comment by ArturDivante** [comment]
On my local (using the same magento), it's not gonna work the same way. Are you testing it for logged in user or logged out? 

I'm asking, because I managed to get any access to country-specific payment method only in one case: I was logged in with all adresses set up in Poland. And payment method setup to polad as well. In this scenario, I did have an access to this method BUT - it was still there, even when i changed county in checkout. For me, atm it looks like logged in user country is checked on this, but any changes of it - not. Like there was no real reaction for country change in checkout - request is sent, but response looks the same...

**Comment by rain2o** [comment]
@ArturDivante I was just testing as a guest, not a logged in user, when making sure it still works. There is an event on the Country field in the billing address that calls the sync payment methods action when it changes.. Are you seeing any requests go out when you change the country? If so, is there a `billingAddress` object in the payload of the request?

**Comment by ArturDivante** [comment]
Well, this is how it looks on my end. Steps description (guest user):
1. I set up Bank Transfer Payment Method for 2 countries - Albania and Algeria. 
2. I switch to VSF and change country in payments secton - request are sent, on of them is 204. No payment methods changes on the front
3. I check the response and see, there are only 'global" methods in it
4. I go back to magento and switch this Bant Transfer Payment method to be global
5. I switch to VSF again, and cange country to any, just to trigger another request 
6. I check response and now this method is listed in it, it's also displayed on front. 

Payments
Only thing that wonders me is that one of the two requests has 204 status, but I have no idea if this is related in any way.

**Comment by rain2o** [comment]
@ArturDivante I noticed the two requests as well, but it was doing that previously so I didn't look into why it was making two requests. I think I see why you aren't getting the correct results, but I don't know why it's happening. The requests you are seeing go to `/payment-methods`, however that endpoint won't accept the full address in the body. So instead, if you have any data filled out (minimally a country) on the billing address then the `syncPaymentMethods` action will actually hit the endpoint `/shipping-information` instead. So for some reason your environment isn't using the updated logic. I'm not sure if it's a matter of not having the most recent version of this branch, or perhaps some weird caching issue. But you should be seeing requests to `/shipping-information` instead of `/payment-methods`.

**Comment by ArturDivante** [comment]
Ok, so that's a solid info. I'm pretty sure I pulled latest changes (did it few times today tbh). Tomorrow i'll investigate this with another QA and her local env. If the problem wont reproduce, then it means its just my local issue and we can merge it, then test after merge and deploy to make sure it's rly ok.

In the meantime, maybe you could share your local.json config with me - maybe some not obvious reason for us to have diffrent results is hiding there.

**Comment by rain2o** [comment]
@ArturDivante Ok that sounds good, thanks!

I'm actually just using `default.json` to ensure everything works out of the box with vsf. I don't have a `local.json` right now for this environment.

**Comment by ArturDivante** [comment]
Hello again! We continued testing it, with my fellow QA @alinadivante and still we have the same results on 3 diffrent computers total. 

I was using unmodified default.json, and local.json containing only api url (we are not using local API). 

```
{
   "api": {
    "url": "https://test.storefrontcloud.io/"
  }
}
```
Tried with demo and next endopoints as well.

Behaviour is still the same as described in Alina's original ticket #2878 - country specific payment methods are only available when specified country == default country set in Stores > configuration > general. When i changed default country to Albania, and then set payment method available for that country, then it's visible on the front - but its visible all the time, no matter which country we set in payment window. Don't know where the diffrence is, but it's not just my local issue :(

**Comment by rain2o** [comment]
@ArturDivante  Ok thanks for the update. I'll try to spin up a fresh instance of VSF off this branch and see if I can replicate the issue. Hopefully I can so I can hunt down what I'm missing.

**Comment by alinadivante** [comment]
yes, I can confirm this problem. If you use default.json, there should be no difference between us, but it works for me in the same way as Arthur wrote above.. I have no idea why

**Comment by rain2o** [comment]
@ArturDivante @alinadivante I just started with a fresh installation of VSF. Cloned the repo, checked out develop, pulled my branch, then ran the yarn installer and that's it, using the demo backend and everything. It is working as I originally described and as expected - It updates based on the billing address country. I'm not really sure at this point what the difference is.

**Comment by andrzejewsky** [comment]
guys it does work! it's probably a question of selected countries, eg. I set UK in shipping and US in the payment and new options appeared. (used `test.storefrontcloud`). @alinadivante @ArturDivante please try with that.

**Comment by alinadivante** [comment]
I have no idea why it works for you and it doesn't work for us. I have "disabled" cache, but maybe problems with changing payment methods result from something caching..

**Comment by rain2o** [comment]
@andrzejewsky Sorry for being non-responsive the past little bit. Thanks for pushing forward without me on this!

**Comment by andrzejewsky** [comment]
@rain2o no problem!

**Comment by ArturDivante** [comment]
@andrzejewsky Could you check if it's correctly deployed and works on test.storefroncloud.io? Because for me behavior on test is exaclty the same. In shipping section, methods are only returned if they are set for 'all countries'. What is more, country selection in "payment' sends and returns no requests at all. I tested it on 2 diffrent computers, all available browsers - inluding freshly installed Opera. If it is some cache issue, then it's deeper than regular browser cache. Im out of ideas here. Pls check it out.

**Pull request merged**